### ISSUE TITLE
Stereo tracks should not allow different sample rates in the channels

### ISSUE THREAD

**Issue by Paul-Licameli** [audacity/audacity#4737]
### Bug description

Mono tracks with different rates can be joined into a stereo track, then split again, retaining their original rates.

This is incompatible with the restructuring needed for wide clips.  An attempt to join two such tracks should make an error message in 3.4.  In the unlikely event that a 3.3 project was saved with such a pathological stereo track, Audacity should refuse to open it.

If you really want to join two channels with mismatched rates, you should resample one or both to a common rate.

### Steps to reproduce

1. Generate some noise in a mono track with default settings; observe "44100 Hz" in the track control panel.
2. Generate another mono track of noise, and use Track > Resample to change it t9 48000 Hz; observe change in its panel.  (observe that it becomes "hairy" when you resample)
3. Select "Make stereo track" from the first track's menu and observe "44100 Hz" in the panel.  (But internally the right channel has not forgotten its original rate.)
4. Now "Split Stereo track" and observe that the second still has 48000 Hz.

But:  also observe that dragging clips into other tracks with different rates is not a problem.  For instance:
- Make a mono noise track at 44100 Hz
- Make new empty monon track; change to 48000 Hz
- Drag the clip from the first to the second
- Observe that it becomes "hairy" indicating that resampling really happened, as it should, to keep new clips in the track consistent in rate

Or you can do this:
- make a problematic track with empty left at 44100, empty right at 48000
- Make new stereo track of noise at 44100, 48000, or some other, higher rate
- Drag stereo clip into the first track
- Observe that respectively right only, left only, or both get "hairy" -- as expected, to keep consistency of clips in each channel

### Expected behavior

Step 3 should fail instead with an error message (what exactly?).

### Actual behavior

See discussion in Steps

### Audacity Version

current master / alpha / nightly build

### Operating system

Windows, macOS, Linux

**Comment by petersampsonaudacity** [comment]
Fascinating catch @Paul-Licameli  - regression testing shows that this is _**NOT**_ a regression - this behavior has obtained all the way from Audacity 1.0.0 through to 3.3.2

I have no idea whether or not Dominic and Roger intended this to be the case or if it was just never blocked _(maybe it never occurred to them that folk might try to do this)_ ...

### Aside
BTW while it is relatively easy now to observe this behavior as one can readily see the tracks rate in its TCP, note that this may not be the case in the future as one of the design Figmas that I have looked at recently proposes a cleaned-up/trimmed TCP that omits the tracks sample rate amongst other useful info. 

![image](https://github.com/audacity/audacity/assets/50205984/b0ac678c-5457-4aa1-9600-19750bf69927)

**Comment by petersampsonaudacity** [comment]
Testing on W10 with @Paul-Licameli 's branch build:  audacity-win-3.4.0-alpha-20230526+4a8eda0-x64

I confirm that the behavior is now blocked:
![image](https://github.com/audacity/audacity/assets/50205984/ff43a2e2-426b-4835-80e2-a0590d1a0f53)

**Track Format**
But note carefully that this fix does not block the creation of a stereo track from two single tracks each with a different **Sample Format**.

I tried with a 16-bit track and a 32-bit track - they could be merged into a single stereo track at 16-bit (the 16-b-t track was the upper track).  And when split again they both reverted to their original bit formats 16-bit and 32-bit.

**Comment by petersampsonaudacity** [comment]
See this PR for further discussion #4739

**Comment by petersampsonaudacity** [comment]
Testing on W10 the use case of "attempting to load a project (made previously) with mismatching rates in a stereo track"

The error message I get is:
![image](https://github.com/audacity/audacity/assets/50205984/26aa0577-655f-4809-996e-4a94a9a090ef)

This message is not helpful at all to the user:

**A) they would need to guess that they could manually fix this:**
1. open it in an earlier version of Audacity
2.  examine each stereo track in turn by splitting to mono, examining the TCP
3. resample requisite tracks to match sample rates
4. rejoin the mono tracks as stereo tracks
5. save the project
6. reopen project in latest Audacity

**B) Audacity offers no repair help**

Most folk are going to struggle to guess or even implement the A) manual fix.

So while we seem to have agreed we should just block such mismatched stereo track creation in 3.4.0 onwards with no automatic resampling - I think a _**good**_ case can be made for offering automatic resampling of channels in stereo tracks as part of the above error message.

**Comment by Tantacrul** [comment]
So after discussing this with the team, we think this PR is good to go in terms of functionality.

However, we need some copy improvements for the dialogs.

Dialog 1: **Mono tracks must have the same sample rate in order to be combined into a stereo track**

Dialog 2: **This project cannot be opened because it contains a stereo track with different sample rates. This functionality is no longer supported since the release of Audacity version 3.4**

Secondly, we'll discuss a potential solution for more gracefully handling the opening of old projects (with incompatible sample rates) in a separate PR. This allows us to close this PR and address any further issues in isolation.

**Comment by petersampsonaudacity** [comment]
@Tantacrul  there remains the parallel issue if mismatched Sample Formats.

Do we need a separate issued logged to track that

**Comment by Tantacrul** [comment]
@petersampsonaudacity: I would have thought so. @crsib - do we think this should be dealt with in a separate PR?

**Comment by Paul-Licameli** [comment]
Joining tracks with mismatched sample formats is a weird thing to do, not obviously useful, and perhaps something you should be warned about in the unlikely vent you make that mistake.

But ruling out that possibility is not of the same importance as for sample rate, to unblock other important developments for this release.

It might be made an issue but not if high priority.

**Comment by petersampsonaudacity** [comment]
See:  Stereo tracks should not allow different sample formats in the channels #4752

**Comment by Ryan-Miller-Audacity** [comment]
Tested and confirmed the updated text for the respective error messages as well as detection of invalid stereo tracks/behaviors.

I did notice two issues with the current implementation: 

1) If you have a stereo track (natively recorded not joined mono tracks, joined 2 mono tracks etc.) and update the sample rate for the track, save the project and attempt to reopen, the different sample rates detector will throw the error message even though there was no misuse/mismatching of rates on the tracks.

Repro steps: 
1. Open Audacity and Record/Generate audio into a stereo track
2. Right click in the TCP and edit the rate of the track 
3. Save the project and exit Audacity
4. Open Audacity and attempt to open your previously saved file
5. Note the mismatch rates error message even though all that has been done is the rate change for a stereo track

2) When you attempt to open a previously created invalid file (with mismatched sample rates), the error will throw; however, upon clicking "OK" on the dialog the project window is not closed and still opens the file but doesn't include the tracks, so it is possible for someone to save their project as an empty project by mistake and lose their previously created work. Given that the nature of opening the invalid file in and of itself is already an edge case, this is really an edge case to an edge case.
 
Repro steps: 
1. Open Audacity and attempt to open an invalid file (you can create one using 3.3.2)
2. Click OK on the error dialog
3. Note that the header of the program shows your file name
4. Save the file
5. Open Audacity 3.3.2 (where you can still open the invalid file) 
6. Note the file is now empty since you saved it with no tracks after the error in 3.4

Note: If you have previously worked within the project window and opening the file would result in a new Audacity window as well, this is not an issue because the OK button closes out the new window.

The first issue above should certainly be fixed. The second issue could be left for further improvements down the road IMO.

**Comment by petersampsonaudacity** [comment]
@Paul-Licameli  -  I suspect that this issue recently reported n the Forum is related to this current one:

**Cannot vocode audio because stereo track does not match.**
https://forum.audacityteam.org/t/cannot-vocode-audio-because-stereo-track-does-not-match-help/76869/1

_Might other effects be similarly affected ?_

**Comment by Ryan-Miller-Audacity** [comment]
Testing the two issues noted in the comment above from previous testing (https://github.com/audacity/audacity/issues/4737#issuecomment-1570799693), I found that issue 2 has been properly fixed and no longer opens an empty version of a file.

Issue 1 still appears to be present as I could not change the rate of a stereo track, save and exit, and then reopen the same file due to the mismatched rates indicator message displaying improperly. It appears this needs further work.

**Comment by petersampsonaudacity** [comment]
Testing on W10 with @Paul-Licameli 's branch:  audacity-win-3.4.0-alpha-20230607+5830faf-x64

### 1. Making a mismatched stereo track
This remains blocked by the revised error message:
![image](https://github.com/audacity/audacity/assets/50205984/6f684fa1-814a-404e-8d13-b867bfaa19bb)

### 2. Spitting a stereo track
Splitting to stereo and mono remain different with 
a) split to stereo producing panned-Left & panned-Right tracks
b) split to mono producing two center-panned tracks

### 3. Project opening with a mismatched stereo track
A project made and saved in 3.3.2 with a stereo track with mis-matched sample rates cannot be opened in Paul's branch:
![image](https://github.com/audacity/audacity/assets/50205984/5e45f338-34d1-465b-ba71-3ba026387671)


### 4. Ryan's issue
@Ryan-Miller-Audacity reported in #4737 

>If you have a stereo track (natively recorded not joined mono tracks, joined 2 mono tracks etc.) and update the sample rate for the track, save the project and attempt to reopen, the different sample rates detector will throw the error message even though there was no misuse/mismatching of rates on the tracks.

I confirm that Paul's new branch also fixes this issue - the project reopens correctly:
a) with no error message
b) with the stereo track at the changed sample rate

**Comment by Ryan-Miller-Audacity** [comment]
Can you check a situation for me in your most recent build for #4739 (https://github.com/audacity/audacity/actions/runs/5204367354?pr=4739)? I believe I have found an issue introduced from the previous round of updates (with the saving fix - https://github.com/audacity/audacity/actions/runs/5144295987).

I am finding that opening a project in the build is not including (or at least not displaying) audio data in the track. 

For example, I created a new save file in 3.3.2 (or the current build; whatever, it doesn't seem to make a difference), and then opened the file in either the most recent or previous build from PR#4739 and the waveform will not display in the track area; however, if I open the same file in the initial build from PR #4739 (https://github.com/audacity/audacity/actions/runs/5142804123) which doesn't include the fix for Issue 2 that was reported previously, the project and its audio data displays as expected.

@petersampsonaudacity could you check and see if you see any similar behavior? I've checked on my Macbook and Windows Laptop (even with clean config just in case) and got the same result in both environments.

**Comment by petersampsonaudacity** [comment]
> @petersampsonaudacity could you check and see if you see any similar behavior? I've checked on my Macbook and Windows Laptop (even with clean config just in case) and got the same result in both environments.

1. I created a project in 3-3-3
2. Saved it
3. closed Audacity
4. Launched build from "most recent build for https://github.com/audacity/audacity/pull/4739 (https://github.com/audacity/audacity/actions/runs/5204367354?pr=4739)"
5. Opened the project
6. Observe: waveforms shown properly

So I cannot seem to reproduce this

**Comment by Paul-Licameli** [comment]
> > @petersampsonaudacity could you check and see if you see any similar behavior? I've checked on my Macbook and Windows Laptop (even with clean config just in case) and got the same result in both environments.
> 
> @Ryan-Miller-Audacity - testing on W10
> 
> 1. I created a project in 3-3-3
> 2. Saved it
> 3. closed Audacity
> 4. Launched build from "most recent build for [Disallow mismatched sample rates #4739](https://github.com/audacity/audacity/pull/4739) (https://github.com/audacity/audacity/actions/runs/5204367354?pr=4739)"
> 5. Opened the project
> 6. Observe: waveforms shown properly
> 
> So I cannot seem to reproduce this
> 
> I also tested opening the 3.3.3 project in the latest 3.4.0 alpha: audacity-win-3.4.0-alpha-20230608+ab110a4-x64 and there too the waveforms are displayed correctly

I did reproduce the problem.  @petersampsonaudacity are you certain you used the correct latest build?

**Comment by Paul-Licameli** [comment]
> Repro steps:
> 
> 1. Open Audacity and attempt to open an invalid file (you can create one using 3.3.2)
> 2. Click OK on the error dialog
> 3. Note that the header of the program shows your file name
> 4. Save the file
> 5. Open Audacity 3.3.2 (where you can still open the invalid file)
> 6. Note the file is now empty since you saved it with no tracks after the error in 3.4

Reminding us of @Ryan-Miller-Audacity 's steps to reproduce another problem when opening an invalid file.

My extra effort to make this case work -- messed up something else.  So I still have work.

I get the problem as simply as this:

1. start Audacity and record some
2. save to a new file
3. close the project
4. (Restart Audacity unless it's Mac)
5. File > Open Recent and choose the recent file
6. See correct duration of the track, but it's silence

**Comment by Paul-Licameli** [comment]
I can also say, the problem is all in the opening, not the saving.  The data are really in the file, and a build without these changes can open it correctly.

**Comment by petersampsonaudacity** [comment]
> I did reproduce the problem. @petersampsonaudacity are you certain you used the correct latest build?

I just tested again with the very latest 3.4.0 alpha build @Paul-Licameli 's build for "Rewrite Paste":
audacity-win-3.4.0-alpha-20230609+cc1d153-x64

And yes the waveforms from the previously saved 3.3.3 project are displayed correctly.

**Comment by Ryan-Miller-Audacity** [comment]
Confirmed final bug fix and initial implementation in most recent build of PR #4739. Testing on windows, mac, and linux

**Comment by petersampsonaudacity** [comment]
I tested this on W10 with the latest master:  audacity-win-3.4.0-alpha-20230609+0ec7ab2-x64-msvc2022

This works fine now:

1.  The creation of a stereo track is blocked
a) if one of the tracks has a Rate change from the TCP 
b) if one of the tracks is Resampled with Tracks > Resample


2.  Projects created in earlier Audacity versions with stereo tracks with mismatched channels (created either way) cannot be opened on 3.4.0 alpha.
@LWinterberg  BUT the error message that is shown has a help link to the GitBook Audacity Support errors page, but there is nothing on that page to advise users as to how they can correct this situation. 

![image](https://github.com/audacity/audacity/assets/50205984/ee424d04-d92a-4885-a030-ce7177699e90)

Links to this page: https://support.audacityteam.org/troubleshooting/error-codes

And note that the error message in the image above does not actually have an "error code" as such.

**Comment by petersampsonaudacity** [comment]
@Paul-Licameli  - having learned yesterday the subtle difference between:
a) rate change of a track from the TCP dropdown menu (where track length changes - and no. of samples stays the same)
b) Reampling of a track with Tracks>Resample (where the track remains the same length but samples are added or removed)

I re-tested using both these methods on one track of a pair of mono tracks split from a stereo track and found:
i) I was unable to make the pair into a stereo track
ii) if I used 3.3.3 or earlier to make such a mismatched stereo track by either method I could not open the project in 3.4.0 alpha
In all four use cases I got an appropriate error message.

**Comment by saintmatthieu** [comment]
Quoting the bug description:
> * Make a mono noise track at 44100 Hz
> * Make new empty monon track; change to 48000 Hz
> * Drag the clip from the first to the second

If I follow these steps now, I get 
![image](https://github.com/audacity/audacity/assets/22740106/72c80063-6b24-4809-8a69-17345ac943b1)

`WaveTrack::mLegacyRate` is uninitialized when `SetRate(mLegacyRate);` gets executed.

**Comment by petersampsonaudacity** [comment]
@saintmatthieu 

Testing on W10 with latest 3.4.0 alpha

**Comment by petersampsonaudacity** [comment]
I added my test results to #4808
